### PR TITLE
Add proper caching to LocaleDataMap

### LINKED ISSUE #16981: 2024.4 Alphas are spamming logs with symbol files not found
### Steps to reproduce:

Run a recent alpha. Try to find an error or anything else in the log (debug level).

### Actual behavior:

Notice that it's difficult to find anything, because the log is full and filling, of tracebacks related to files not found.

```
IO - speech.speech.speak (04:38:17.638) - MainThread (25064):
Speaking ['alpha', '12 of 69', CancellableSpeech (still valid)]
DEBUGWARNING - characterProcessing._getSpeechSymbolsForLocale (04:38:17.638) - MainThread (25064):
Error loading 'cldr' symbols for locale 'en_GB'
Traceback (most recent call last):
  File "characterProcessing.pyc", line 414, in _getSpeechSymbolsForLocale
  File "characterProcessing.pyc", line 818, in getSymbols
  File "characterProcessing.pyc", line 64, in fetchLocaleData
  File "characterProcessing.pyc", line 826, in _initSymbols
FileNotFoundError: No 'cldr' data for locale 'en_GB'
DEBUGWARNING - characterProcessing._getSpeechSymbolsForLocale (04:38:17.644) - MainThread (25064):
Error loading 'builtin' symbols for locale 'en_GB'
Traceback (most recent call last):
  File "characterProcessing.pyc", line 414, in _getSpeechSymbolsForLocale
  File "characterProcessing.pyc", line 818, in getSymbols
  File "characterProcessing.pyc", line 64, in fetchLocaleData
  File "characterProcessing.pyc", line 826, in _initSymbols
FileNotFoundError: No 'builtin' data for locale 'en_GB'
DEBUGWARNING - characterProcessing._getSpeechSymbolsForLocale (04:38:17.644) - MainThread (25064):
Error loading 'cldr' symbols for locale 'en_GB'
Traceback (most recent call last):
  File "characterProcessing.pyc", line 414, in _getSpeechSymbolsForLocale
  File "characterProcessing.pyc", line 818, in getSymbols
  File "characterProcessing.pyc", line 64, in fetchLocaleData
  File "characterProcessing.pyc", line 826, in _initSymbols
FileNotFoundError: No 'cldr' data for locale 'en_GB'
```
(And again, and again, and again...)

### Expected behavior:

Maybe one notice about files not found the first time the locale is used, and then no more.
I'm also concerned it's having an effect of performance, to be polling for these files two or three times every utterance, if indeed that's what's happening.

### NVDA logs, crash dumps and other attachments:

[nvda-logSpam.txt](https://github.com/user-attachments/files/16540439/nvda-logSpam.txt)

#### NVDA installed/portable/running from source:

Portable

#### NVDA version:

alpha-33301,b8de34e9

#### Windows version:

11 Home 23H3

#### Name and version of other software in use when reproducing the issue:

N/A

#### Does the issue still occur after restarting your computer?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

Yes. 2024.3beta6 (installed), is not doing this.

#### If NVDA add-ons are disabled, is your problem still occurring?

Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?

Yes

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #16981 
Follow up of #16823

### Summary of the issue:
Since #16823, the log is spammed with errors re loading of nonexistent symbol dictionaries.

### Description of user facing changes
No spamming with errors in the log.

### Description of development approach
Before #16823, there were sets for caching locales without symbol dictionaries. When no dictionaries existed, the locales were saved in the cache and upon next request, `_getSpeechSymbolsForLocale` returned early.
That caching mechanism was apparently removed in #16823. I now restored similar caching, but the cache is now stored on the `LocaleDataMap` instance. This should also be beneficial for character descriptions.

### Testing strategy:
Ensured that the error @XLTechie had only occurs once per dictionary.

### Known issues with pull request:
None known

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
	- Enhanced locale data fetching efficiency by introducing a cache for locales with no associated data.
  
- **Bug Fixes**
	- Modified methods to ensure proper management of the cache when locale data is invalidated or cleared.

### CODE REVIEW

**Comment by XLTechie** [comment]
Thanks for such quick work, @LeonarddeR!

File: source/characterProcessing.py [hunk 1]
```diff
@@ -34,6 +34,8 @@
 class LocaleDataMap(Generic[_LocaleDataT], object):
 	"""Allows access to locale-specific data objects, dynamically loading them if needed on request"""
 
+	_noDataLocalesCache: set[str]
```
Comment (seanbudd) [inline source/characterProcessing.py:37]: is there a reason why this was added to the class rather than kept in `__init__`?

**Review by SaschaCowley — APPROVED** [review]

File: source/characterProcessing.py (hunk 1 above)
Comment (LeonarddeR) [reply source/characterProcessing.py:37]: This is only a type hint, the set itself is initiated in `__init__`

Comment (seanbudd) [reply source/characterProcessing.py:37]: normally the type hint would go in `__init__` though right?

**Review by SaschaCowley — APPROVED** [review]

**Pull request merged**